### PR TITLE
CB-11572 add datalakecrn to stack in order to eliminate datalake resource; all datalakeresource activity moved to datalakeservice

### PR DESCRIPTION
**Pull request by afarsang**

### CODE REVIEW

File: core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
+-- // CB-11572 datahub connected to the datalake without datalakeresource
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE stack ADD COLUMN IF NOT EXISTS datalakecrn varchar(512);
+
+CREATE INDEX idx_stack_datalakecrn ON stack (datalakecrn);
+
+UPDATE stack s
```
Comment (bergerdenes) [inline core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql:8]: this update needs to be repeated after this PR is on PROD!

File: core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+SET datalakecrn = dl.resourcecrn
+FROM (
+  SELECT s1.id, s2.resourcecrn FROM stack s1, datalakeresources d, stack s2 WHERE s1.type = 'WORKLOAD' AND s1.datalakeresourceid = d.id AND d.datalakestack_id = s2.id
```
Comment (bergerdenes) [inline core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql:11]: JOIN ... ON

File: core/src/test/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.Optional;
+
+import org.junit.Before;
+import org.junit.Test;
```
Comment (bergerdenes) [inline core/src/test/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeServiceTest.java:13]: Please use JUnit5. And for all other test classes as well.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java [hunk 4]
```diff
@@ -86,6 +99,7 @@ private boolean decorateStackWithConfigs(Stack publicStack, Optional<DatalakeRes
         if (datalakeResource.isPresent()) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java:99]: How will this be removed in the next phase?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+import com.sequenceiq.cloudbreak.domain.stack.cluster.DatalakeResources;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ServiceDescriptor;
+import com.sequenceiq.cloudbreak.domain.view.ClusterApiView;
+import com.sequenceiq.cloudbreak.service.datalake.DatalakeResourcesService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.template.views.SharedServiceConfigsView;
+import com.sequenceiq.cloudbreak.workspace.model.Workspace;
+
+@Service
+public class DatalakeService {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:41]: There are a few methods that lack logging.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+            Optional<DatalakeResources> datalakeResource = datalakeResourcesService.findById(cluster.getStack().getDatalakeId());
+            if (datalakeResource.isPresent()) {
+                sharedServiceResponse.setSharedClusterId(datalakeResource.get().getDatalakeStackId());
+                sharedServiceResponse.setSharedClusterName(datalakeResource.get().getName());
+            }
+        }
+        clusterResponse.setSharedServiceResponse(sharedServiceResponse);
+    }
+
+    public void addSharedServiceResponse(Stack stack, StackV4Response stackResponse) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:101]: This method is too complex

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        Long datalakeResourceId = datahubStack.getDatalakeResourceId();
+        if (datalakeResourceId != null) {
+            Optional<DatalakeResources> datalakeResource = datalakeResourcesService.findById(datalakeResourceId);
+            if (datalakeResource.isPresent() && datalakeResource.get().getDatalakeStackId() != null) {
+                Long datalakeStackId = datalakeResource.get().getDatalakeStackId();
+                Stack datalakeStack = stackService.getByIdWithListsInTransaction(datalakeStackId);
+                return Optional.of(datalakeStack);
+            }
+        }
+        LOGGER.info("There is no datalake resource id has set for the cluster.");
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:156]: -nor datalake crn-

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java (hunk 6 above)
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:101]: this is how I found it elsewhere

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java [hunk 8]
```diff
@@ -40,10 +43,20 @@
...
+    @Inject
+    private RdsConfigService rdsConfigService;
+
     public Cluster configureCluster(@Nonnull Cluster requestedCluster, User user, Workspace workspace) {
         Objects.requireNonNull(requestedCluster);
         Stack stack = requestedCluster.getStack();
-        if (stack.getDatalakeResourceId() != null) {
+        if (!Strings.isNullOrEmpty(stack.getDatalakeCrn())) {
+            Stack datalakeStack = stackService.getByCrn(stack.getDatalakeCrn());
+            if (stack != null) {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java:54]: nice catch

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java [hunk 9]
```diff
@@ -328,8 +335,27 @@ public StackClusterStatusView getStatusByNameOrCrn(NameOrCrn nameOrCrn, Long wor
...
-        return datalakeResources.isEmpty() ? Collections.emptySet() : stackRepository.findEphemeralClusters(datalakeResources.get().getId());
+        Stack datalakeStack = get(datalakeStackId);
+        Set<StackIdView> stacksConnectedByDatalakeCrn;
+        if (datalakeStack != null) {
+            stacksConnectedByDatalakeCrn = findByDatalakeCrn(datalakeStack.getResourceCrn());
+        } else {
+            stacksConnectedByDatalakeCrn = Collections.emptySet();
+        }
+
+        Optional<Long> id = datalakeService.getDatalakeResourceId(datalakeStackId);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java:346]: I guess I'm right:
accidentally stack could be connected through datalakeresource only (created before the release, but not migrated - may happen)
or
stack could be connected through datalakecrn only (created after the release)
or 
both (created before the release, and migrated).

File: core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql (hunk 1 above)
Comment (afarsang) [reply core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql:8]: this is also in the cb12835

File: core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql (hunk 2 above)
Comment (afarsang) [reply core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql:11]: this is equivalent

File: core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToStackConverter.java [hunk 10]
```diff
@@ -159,7 +157,8 @@ public Stack convert(StackV4Request source) {
         setTimeToLive(source, stack);
         stack.setWorkspace(workspace);
         stack.setDisplayName(source.getName());
-        stack.setDatalakeResourceId(getDatalakeResourceId(source, workspace));
+        datalakeService.setDatalakeIdOnStack(stack, source, workspace);
+        stack.setDatalakeCrn(datalakeService.getDatalakeCrn(source, workspace));
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackV4RequestToStackConverter.java:161]: nope. thx

File: core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java [hunk 11]
```diff
@@ -557,17 +554,13 @@ private void saveCustomNameservers(Stack stack, KerberosConfig kerberosConfig, M
...
-            Optional<DatalakeResources> datalakeResource = datalakeResourcesService.findById(datalakeResourceId);
-            if (datalakeResource.isPresent() && datalakeResource.get().getDatalakeStackId() != null) {
-                Long datalakeStackId = datalakeResource.get().getDatalakeStackId();
-                Stack dataLakeStack = stackService.getByIdWithListsInTransaction(datalakeStackId);
-                String datalakeDomain = dataLakeStack.getGatewayInstanceMetadata().get(0).getDomain();
-                List<String> ipList = dataLakeStack.getGatewayInstanceMetadata().stream().map(InstanceMetaData::getPrivateIp).collect(Collectors.toList());
-                servicePillar.put("forwarder-zones", new SaltPillarProperties("/unbound/forwarders.sls",
-                        singletonMap("forwarder-zones", singletonMap(datalakeDomain, singletonMap("nameservers", ipList)))));
-            }
+        Optional<Stack> datahubStackOptional = datalakeService.getDatalakeStackByDatahubStack(stack);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/ClusterHostServiceRunner.java:557]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java (hunk 4 above)
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java:99]: https://github.com/hortonworks/cloudbreak/blob/cb12835/core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java

like this:
`
    public Cluster configureCluster(@Nonnull Cluster requestedCluster, User user, Workspace workspace) {
      
      
                Objects.requireNonNull(requestedCluster);
      
      
                Stack stack = requestedCluster.getStack();
      
      
                if (!Strings.isNullOrEmpty(stack.getDatalakeCrn())) {
      
      
                    Stack datalakeStack = stackService.getByCrn(stack.getDatalakeCrn());
      
      
                    if (stack != null) {
      
      
                        Set&lt;RDSConfig&gt; rdsConfigs = rdsConfigService.findByClusterId(datalakeStack.getCluster().getId());
      
      
                        setupRds(requestedCluster, rdsConfigs);
      
      
                        setupStoragePath(requestedCluster, datalakeStack);
      
      
                    }
      
      
                }
      
      
                return requestedCluster;
      
      
            }`

Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/SharedServiceConfigProvider.java:99]: I am not sure understand completely the question. I referenced the  next phase branch

File: core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql [hunk 12]
```diff
@@ -0,0 +1,17 @@
+-- // CB-11572 datahub connected to the datalake without datalakeresource
+-- Migration SQL that makes the change goes here.
+
+ALTER TABLE stack ADD COLUMN IF NOT EXISTS datalakecrn varchar(512);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/resources/schema/app/20210611131901_CB-11572_datahub_connected_to_datalake_without_datalakeresource.sql:4]: As I recall Ricsi asked me to do so (there is another place to where 512).

File: core/src/test/java/com/sequenceiq/cloudbreak/service/stack/StackIdViewImpl.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
+package com.sequenceiq.cloudbreak.service.stack;
+
+
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/test/java/com/sequenceiq/cloudbreak/service/stack/StackIdViewImpl.java:3]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackService.java:346]: you are right (except, there is no such case as "stack could be connected through datalakecrn only"), the 2 cases must be handled

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+
+    private Optional<DatalakeResources> getDatalakeResourceId(StackV4Request source, Workspace workspace) {
+        try {
+            if (source.getSharedService() != null && isNotBlank(source.getSharedService().getDatalakeName())) {
+                return Optional.of(datalakeResourcesService.getByNameForWorkspace(source.getSharedService().getDatalakeName(), workspace));
+            }
+        } catch (NotFoundException nfe) {
+            LOGGER.debug("No datalake resource found for data lake: {}", source.getSharedService().getDatalakeName());
+        }
+        return Optional.empty();
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:267]: fixed

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+
+        return sharedServiceConfigsView;
+    }
+
+    private String getDatalakeClusterManagerFqdn(Stack datalakeStack) {
+        String fqdn;
+        try {
+            fqdn = transactionService.required(() -> datalakeStack.getGatewayInstanceMetadata().isEmpty()
+                    ? datalakeStack.getClusterManagerIp() : datalakeStack.getGatewayInstanceMetadata().iterator().next().getDiscoveryFQDN());
+        } catch (Exception nothing) {
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:321]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+            if (source.getSharedService() != null && isNotBlank(source.getSharedService().getDatalakeName())) {
+                return Optional.of(datalakeResourcesService.getByNameForWorkspace(source.getSharedService().getDatalakeName(), workspace));
+            }
+        } catch (NotFoundException nfe) {
+            LOGGER.debug("No datalake resource found for data lake: {}", source.getSharedService().getDatalakeName());
+        }
+        return Optional.empty();
+    }
+
+    private SharedServiceConfigsView createSharedServiceConfigs(Blueprint blueprint, String ambariPassword,
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:270]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+            Optional<DatalakeResources> datalakeResources) {
+        SharedServiceConfigsView sharedServiceConfigsView = new SharedServiceConfigsView();
+        if (datalakeResources.isPresent()) {
+            DatalakeResources datalakeResource = datalakeResources.get();
+            String rangerPort = getRangerPort(datalakeResource.getServiceDescriptorMap(), DEFAULT_RANGER_PORT);
+            sharedServiceConfigsView.setRangerAdminPassword(getRangerAdminPassword(datalakeResource).orElse(null));
+            sharedServiceConfigsView.setAttachedCluster(true);
+            sharedServiceConfigsView.setDatalakeCluster(false);
+            sharedServiceConfigsView.setDatalakeClusterManagerIp(datalakeResource.getDatalakeAmbariIp());
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:279]: yes. this is how I found it, and below what I wrote for datalake crn. I have also found that most of the fields in sharedserviceconfigview is not used

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+            sharedServiceConfigsView.setDatalakeClusterManagerFqdn(datalakeResource.getDatalakeAmbariFqdn());
+            sharedServiceConfigsView.setDatalakeComponents(datalakeResource.getDatalakeComponentSet());
+            sharedServiceConfigsView.setRangerAdminPort(rangerPort);
+            Optional<String> rangerAdminHost = getHostForComponentInService(datalakeResource, RANGER, RANGER_ADMIN);
+            sharedServiceConfigsView.setRangerAdminHost(rangerAdminHost.orElse(null));
+        } else if (blueprintUtils.isSharedServiceReadyBlueprint(blueprint)) {
+            sharedServiceConfigsView.setRangerAdminPassword(ambariPassword);
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:286]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+
+    public Optional<Stack> getDatalakeStackByStackEnvironmentCrn(Stack datahubStack) {
+        if (StackType.DATALAKE.equals(datahubStack.getType())) {
+            return Optional.empty();
+        }
+        List<StackStatusView> res = stackService.getByEnvironmentCrnAndStackType(datahubStack.getEnvironmentCrn(), StackType.DATALAKE);
+        if (res.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(stackService.getByIdWithListsInTransaction(res.get(0).getId()));
```
[the remark this answers is not in the document]
Comment (afarsang) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/sharedservice/DatalakeService.java:170]: commented

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**